File: 1084-find-k-length-substrings-with-no-repeated-characters/find-k-length-substrings-with-no-repeated-characters.java
class Solution {
    public int numKLenSubstrNoRepeats(String s, int k) {
        if (k > 26 || k < 1 || s.length() < k) {
            return 0;
        }

        HashSet<Character> charSet = new HashSet<>();
        int ans = 0;
        int left = 0;

        for (int right = 0; right < s.length(); right++) {
            while (charSet.contains(s.charAt(right))) {
                charSet.remove(s.charAt(left));
                left++;
            }

            charSet.add(s.charAt(right));

            ans = right - left + 1 >= k ? ans + 1 : ans;
        }

        return ans;
    }
}

File: 1084-find-k-length-substrings-with-no-repeated-characters/find-k-length-substrings-with-no-repeated-characters.py
class Solution:
    def numKLenSubstrNoRepeats(self, s: str, k: int) -> int:
        substrings = {}
        left = 0
        right = k
        ans = 0

        while right <= len(s):
            substring = s[left:right]
            if substring in substrings:
                substrings[substring] += 1
            else:
                substrings[substring] = 1
            left += 1
            right += 1

        for substring in substrings:
            if (len(substring) == len(set(substring))):
                ans += substrings[substring]
        
        return ans
